### PR TITLE
WIP Add support for hostname validation #23565

### PR DESCRIPTION
**Pull request by mnordie**

### CODE REVIEW

**Comment by raboof** [comment]
OK TO TEST

File: akka-remote/src/main/scala/akka/remote/transport/netty/NettyTransport.scala [hunk 1]
```diff
@@ -361,7 +361,7 @@ class NettyTransport(val settings: NettyTransportSettings, val system: ExtendedA
     new ChannelPipelineFactory {
       override def getPipeline: ChannelPipeline = {
         val pipeline = newPipeline
-        if (EnableSsl) pipeline.addFirst("SslHandler", sslHandler(isClient = true))
+        if (EnableSsl) pipeline.addFirst("SslHandler", sslHandler(Option(remoteAddress), isClient = true))
```
Comment (raboof) [inline akka-remote/src/main/scala/akka/remote/transport/netty/NettyTransport.scala]: Is `remoteAddress` nullable here? Doesn't *seem* so, in that case `Some(remoteAddress)` might be clearer.

File: akka-remote/src/main/scala/akka/remote/transport/netty/NettyTransport.scala [hunk 2]
```diff
@@ -340,16 +340,16 @@ class NettyTransport(val settings: NettyTransportSettings, val system: ExtendedA
 
   private val associationListenerPromise: Promise[AssociationEventListener] = Promise()
 
-  private def sslHandler(isClient: Boolean): SslHandler = {
-    val handler = NettySSLSupport(settings.SslSettings.get, log, isClient)
+  private def sslHandler(remoteAddress: Option[Address], isClient: Boolean): SslHandler = {
```
Comment (raboof) [inline akka-remote/src/main/scala/akka/remote/transport/netty/NettyTransport.scala]: perhaps we should split this in a `sslServerHandler` and `sslClientHandler` to make explicit that a server handler never has a `remoteAddress` and a client always does?

**Review by raboof — APPROVED** [review]
LGTM, some comments for discussion/consideration

File: akka-remote/src/main/resources/reference.conf [hunk 3]
```diff
@@ -669,6 +669,9 @@ akka {
         # The first step ensures that all nodes will send over a certificate when asked to. The second
         # step will ensure that all nodes finally enforce the secure checking of client certificates.
         require-mutual-authentication = on
+
+        # Require hostname validation of peer address against certificate
+        require-hostname-validation = off
```
[the remark this answers is not in the document]
Comment (mnordie) [reply akka-remote/src/main/resources/reference.conf]: I can't really say. Is there somebody else we can get an opinion from? @jrudolph maybe?

[the remark this answers is not in the document]
Comment (jrudolph) [reply akka-remote/src/main/resources/reference.conf]: I don't think we can make that change right now, because up to now it was possible (and maybe common?) to use the same set of keys for all nodes.

**Comment by mnordie** [comment]
@jrudolph what is your take on this? Need anything else?

**Comment by patriknw** [comment]
Refs #23565

File: akka-docs/src/main/paradox/scala/remoting.md [hunk 4]
```diff
@@ -594,7 +594,8 @@ While Akka is a peer-to-peer technology, each connection between nodes starts ou
 the other (the "server").
 
 Note that if TLS is enabled with mutual authentication there is still a risk that an attacker can gain access to a valid certificate
-by compromising any node with certificates issued by the same internal PKI tree.
+by compromising any node with certificates issued by the same internal PKI tree. To limit this risk hostname validation can be enabled.
```
Comment (EronWright) [inline akka-docs/src/main/paradox/scala/remoting.md]: I don't think that hostname verification is improving the situation as much as this edit suggests.   Hostname verification improves the trustworthiness of the passive side but not the active side of the peer-to-peer relationship.    It will still be the case that any client with a certificate issued by the same PKI can connect, since Akka has no authorization logic (to my knowledge) beyond 'is authenticated'.

In other words, the root cause of the issue alluded to by the above security note is a lack of authorization logic in Akka.  Could the note be clarified to that effect?

**Comment by EronWright** [comment]
Given the peer-to-peer nature of Akka and the fact that hostname verification doesn't verify the client's address, I strongly suggest that the security documentation be careful to not overstate the benefits of this patch.   An application-level verification of the certificate on both sides of the connection would be a much stronger feature in my opinion.

**Comment by raboof** [comment]
I agree being more clear/explicit about the exact security implications of those settings is useful.

I'm not sure what kind of "application-level verification" would actually add security though, after all the end result will still not be any more granular than "is authenticated". What would you like to be able to do that you cannot achieve by configuring your key- and truststores as desired?

**Comment by EronWright** [comment]
I've been reflecting on how to characterize Akka's authorization story, and found it helpful to think of the truststore as an access control list.   Any peer certificate that is rooted in the provided truststore is granted access.  To narrow the scope of acceptable certificates, one must carefully construct the truststore.   e.g. simply using your system's truststore would effectively grant the whole internet access to your Akka system.   A similar rationale is used in [RFC-2818](https://tools.ietf.org/search/rfc2818#section-3.1) to justify skipping hostname verification. 

@raboof I agree with your points but I'm calling out that it is unintuitive and inconvenient to craft a sufficiently narrow truststore.   By 'app-level validation' I mean additional checks of the peer certificate in Akka itself, e.g. checking the subjectAltName against a whitelist.   Hostname verification is a partial fix along those lines but not sufficient.

**Comment by EronWright** [comment]
Let me restate my objection, which is that akka lacks _authorization_ logic ('app validation logic' in my earlier comments).   The truststore and hostname validation simply relates to _authentication_, which sets a very low bar if you simply use the system truststore.     Hostname validation does not add much of value to Akka.
